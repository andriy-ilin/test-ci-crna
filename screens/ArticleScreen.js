import React, { Component } from "react";
import { inject, observer } from "mobx-react/native";
import { View, ScrollView, Text, Image, TouchableOpacity } from "react-native";
import { withNamespaces } from "react-i18next";

import Header from "../components/Header";

@withNamespaces(["home", "common"], { wait: true })
@observer
export class ArticleScreen extends Component {
  render() {
    const { t } = this.props;
    return (
      <View>
        <TouchableOpacity
          onPress={() => this.props.navigation.navigate("home")}
        >
          <Text>Back</Text>
        </TouchableOpacity>
        <Text>ArticleScreen</Text>
      </View>
    );
  }
}

export default ArticleScreen;
