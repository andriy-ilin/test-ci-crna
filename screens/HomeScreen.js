import React, { Component } from "react";
import { View, ScrollView, Text, Image, TouchableOpacity } from "react-native";
import { inject, observer } from "mobx-react/native";
import { withNamespaces } from "react-i18next";

import Title from "../components/Title";
import StyledText from "../components/StyledText";
import { AllStoriesContainer } from "./RegionsScreen";

@withNamespaces(["home"], { wait: true })
@inject("article")
@observer
export default class CatalogArticles extends Component {
  state = {
    loaded: false,
    catalogArticles: []
  };

  async componentDidMount() {
    const { article } = this.props;
    await article.getArticle(`/catalog/en`);
    this.setState({ loaded: true, catalogArticles: article.list });
  }

  render() {
    const {
      t,
      article,
      navigation: { navigate }
    } = this.props;
    return (
      <View>
        <Title>{t("Top stories")}</Title>
        <StyledText.Light>
          {t("Discover more than 200 breathtaking articles about Ukraine")}
        </StyledText.Light>
        <AllStoriesContainer
          list={article.list}
          onPress={id => {
            navigate("article", { id });
          }}
        />
      </View>
    );
  }
}

@inject("navigation")
export class Article extends Component {
  render() {
    const { mainBg, mainTitle, id } = this.props;
    return (
      <View>
        <TouchableOpacity onPress={() => this.handleEventPress(id)}>
          <Text>{`${mainTitle}`}</Text>
          <Image
            style={{
              width: 51,
              height: 51
            }}
            source={{ uri: mainBg }}
          />
        </TouchableOpacity>
      </View>
    );
  }
  handleEventPress = ({ id }) => this.props.navigation.goTo("article");
}
