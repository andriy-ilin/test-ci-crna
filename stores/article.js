import { observable, computed, action, toJS } from "mobx";
import api from "../services/api";
import { entitiesFromFB } from "./utils";

import BasicStore from "./BasicStore";

class ArticleStore extends BasicStore {
  @observable loading = false;
  @observable loaded = false;

  @observable entities = {};
  @observable articleData = {};
  @observable regionArticleIdData = {};

  @computed get list() {
    return Object.values(this.entities);
  }

  @computed get article() {
    return toJS(this.articleData);
  }
  @computed get regionArticleId() {
    return toJS(this.regionArticleIdData);
  }

  @action set(name, data) {
    this[name] = data;
  }

  @action
  async getArticle(refName, name = "entities") {
    const data = await api.fetchAllByEntityName(refName);
    this.set(name, data);
  }
}

export default ArticleStore;
