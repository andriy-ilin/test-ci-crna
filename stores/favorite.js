import { observable, computed, action, toJS } from "mobx";
import { AsyncStorage } from "react-native";
import api from "../services/api";
import { entitiesFromFB } from "./utils";

import BasicStore from "./BasicStore";

class FavoriteStore extends BasicStore {
  @observable loading = false;
  @observable loaded = false;

  @observable entities = {};
  @observable favoritesEntities = [];

  @computed get list() {
    return Object.values(this.entities);
  }
  @computed get listFavorites() {
    return Object.values(this.favoritesEntities);
  }

  @action
  async getFavoritesArticles() {
    this.set("loading", true);
    await this.getArticles(`/catalog/en`);
    const favorites = await this.getStorage("@ukrainer:favorites");
    this.set("loading", false);

    return this.set(
      "favoritesEntities",
      this.list.filter(({ id }) => favorites && favorites.includes(id))
    );
  }

  @action set(name, data) {
    this[name] = data;
  }

  @action
  async saveToStorage(id) {
    try {
      const favorites = await this.getStorage("@ukrainer:favorites");
      const list = favorites
        ? JSON.parse(favorites).includes(id)
          ? JSON.parse(favorites)
          : [...JSON.parse(favorites), id]
        : [id];
      await AsyncStorage.setItem("@ukrainer:favorites", JSON.stringify(list));
      this.getFavoritesArticles();
    } catch (error) {
      console.log(">>>----error save storage", error);
    }
  }

  @action
  async getStorage(key) {
    try {
      const value = await AsyncStorage.getItem(key);
      return value;
    } catch (error) {
      console.log(">>>----error read from storage", error);
    }
  }

  @action
  async deleteFromStorage(id) {
    try {
      const favorites = await this.getStorage("@ukrainer:favorites");
      const listFilter = JSON.parse(favorites).filter(el => el !== id);
      await AsyncStorage.setItem(
        "@ukrainer:favorites",
        JSON.stringify(listFilter)
      );

      this.set(
        "favoritesEntities",
        this.list.filter(({ id }) => listFilter && listFilter.includes(id))
      );
    } catch (error) {
      console.log(">>>----error read from storage", error);
    }
  }
  @action
  async clearStorage(key, id) {
    try {
      await AsyncStorage.removeItem("@ukrainer:favorites");
      const favorites = await this.getStorage("@ukrainer:favorites");
      this.set("favoritesEntities", []);
    } catch (error) {
      console.log(">>>----error read from storage", error);
    }
  }

  @action
  async getArticles(refName, name = "entities") {
    this.set("loading", true);
    const data = await api.fetchAllByEntityName(refName);
    this.set(name, data);
    this.set("loading", false);
  }
}

export default FavoriteStore;
