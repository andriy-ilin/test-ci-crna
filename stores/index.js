import "../helpers/firebaseConfig";

import Navigation from "./navigation";
import Article from "./article";

const stores = {};
stores.navigation = new Navigation(stores);
stores.article = new Article(stores);

export default stores;
