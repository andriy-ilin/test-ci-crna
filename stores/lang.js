import { observable, computed, action, toJS } from "mobx";
import { AsyncStorage } from "react-native";
import { Localization } from "expo-localization";
import i18n from "../services/i18n";

import api from "../services/api";
import { entitiesFromFB } from "./utils";

import BasicStore from "./BasicStore";

class LangStore extends BasicStore {
  @observable loading = false;
  @observable loaded = false;
  @observable currentLang = "";
  @observable entities = {};
  @observable translateList = [];

  @computed get list() {
    return Object.values(this.entities);
  }

  @computed get articlesTranslateList() {
    return Object.values(this.entities).map(({ key, value }) => ({
      key,
      value,
      link: this.translateList[key]
    }));
  }

  @action set(name, data) {
    this[name] = data;
  }

  @action setArticlesTranslateList(list) {
    this.set(
      "translateList",
      list.reduce((prev, { lang, href }) => ({ ...prev, [lang]: href }), {})
    );
    return list;
  }

  @action async detectLang() {
    const { locale } = await Localization.getLocalizationAsync();
    const lang = locale.includes("-") ? locale.split("-")[0] : locale;
    return lang;
  }

  @action async getCurrentLangFromStorage() {
    const value = await AsyncStorage.getItem("@ukrainer:lang");
    return value && JSON.parse(value);
  }

  @action
  async setCurrentLangFromStorage() {
    const lang =
      (await this.getCurrentLangFromStorage()) || (await this.detectLang());
    await i18n.changeLanguage(lang);
    await AsyncStorage.setItem("@ukrainer:lang", JSON.stringify(lang));
    this.set("currentLang", lang);
    return;
  }

  @action
  async changeCurrentLang(lang, { changeLanguage = () => {} }) {
    await i18n.changeLanguage(lang);
    await AsyncStorage.setItem("@ukrainer:lang", JSON.stringify(lang));
    this.set("currentLang", lang);
    return;
  }

  @action
  async getCatalog(lang, link) {
    const data = await api.fetchAllByEntityName(`/catalog/${lang}`);
    const [{ id }] = Object.values(data).filter(({ href }) => href === link);
    return id;
  }

  @action
  async getLang(refName, name = "entities") {
    this.set("loading", true);
    const data = await api.fetchAllByEntityName(refName);
    this.set(name, data);
    return this.set("loading", false);
  }
}

export default LangStore;
