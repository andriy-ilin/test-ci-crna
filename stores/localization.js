import { observable, computed, action } from "mobx";
import i18n from "../services/i18n";

import BasicStore from "./BasicStore";

class LocalizationStore extends BasicStore {
  @observable loading = false;
  @observable loaded = false;
  @observable lang = null;

  @action getLocale() {
    return i18n.language;
  }

  @action
  async setLocale(name) {
    await i18n.changeLanguage(name);
    return this.set("lang", name);
  }

  @action set(name, data) {
    return (this[name] = data);
  }
  @action get(name) {
    return this[name];
  }
}

export default LocalizationStore;
