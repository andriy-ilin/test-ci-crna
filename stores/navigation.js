import { autorun } from "mobx";
import { NavigationActions, StackActions } from "react-navigation";

import BasicStore from "./BasicStore";

class NavigationStore extends BasicStore {
  ref = null;

  initStore() {
    let firstRun = true;

    setTimeout(() => {
      autorun(() => {
        const user = null;
        const screen = user ? "home" : "article";

        if (!firstRun) {
          this.goTo(screen);
        }

        firstRun = false;
      });
    });
  }

  setNavRef = ref => {
    this.ref = ref;
    this.initStore();
  };

  goTo = (routeName, params) =>
    this.ref.dispatch(
      NavigationActions.navigate({
        routeName,
        params
      })
    );

  goBack = () => this.ref.dispatch(NavigationActions.back());

  reset = routeName =>
    this.ref.dispatch(
      StackActions.reset({
        index: 0,
        actions: [NavigationActions.navigate({ routeName })]
      })
    );
}

export default NavigationStore;
