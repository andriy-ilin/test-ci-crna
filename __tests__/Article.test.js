import "react-native";
import React from "react";
import { Provider } from "mobx-react/native";
import { mount } from "enzyme";

import ArticleScreen from "../screens/ArticleScreen";
import HomeScreen, { Article } from "../screens/HomeScreen";
import stores from "../stores";

describe("Article snapshot", () => {
  jest.useFakeTimers();

  // it("renders HomeScreen", async () => {
  //   const rendered = renderer.create(<HomeScreen />).toJSON();
  //   expect(rendered).toMatchSnapshot();
  // });

  it("renders ArticleScreen", async () => {
    const rendered = mount(<ArticleScreen />);
    expect(rendered).toMatchSnapshot();
  });

  it("check ArticleScreen text inside", async () => {
    const wrapper = mount(<ArticleScreen />);
    const text = wrapper
      .find("Text")
      .at(1)
      .text();
    expect(text).toEqual("ArticleScreen");
  });

  it("Article snapshot", async () => {
    const wrapper = mount(
      <Provider {...stores}>
        <Article />
      </Provider>
    );

    expect(wrapper).toMatchSnapshot();
  });

  it("Article props navigate what inject from MobX", async () => {
    const wrapper = mount(
      <Provider {...stores}>
        <Article />
      </Provider>
    );
    const {
      navigation: { goTo }
    } = wrapper.props();
    expect(goTo).toBeTruthy();
  });
});
