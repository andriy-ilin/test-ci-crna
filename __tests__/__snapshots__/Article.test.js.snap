// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Article snapshot Article snapshot 1`] = `
<Provider
  article={
    ArticleStore {
      "allStores": Object {
        "article": [Circular],
        "navigation": NavigationStore {
          "allStores": [Circular],
          "goBack": [Function],
          "goTo": [Function],
          "ref": null,
          "reset": [Function],
          "setNavRef": [Function],
        },
      },
    }
  }
  navigation={
    NavigationStore {
      "allStores": Object {
        "article": ArticleStore {
          "allStores": [Circular],
        },
        "navigation": [Circular],
      },
      "goBack": [Function],
      "goTo": [Function],
      "ref": null,
      "reset": [Function],
      "setNavRef": [Function],
    }
  }
>
  <inject-Article-with-navigation>
    <Article
      navigation={
        NavigationStore {
          "allStores": Object {
            "article": ArticleStore {
              "allStores": [Circular],
            },
            "navigation": [Circular],
          },
          "goBack": [Function],
          "goTo": [Function],
          "ref": null,
          "reset": [Function],
          "setNavRef": [Function],
        }
      }
    >
      <Component>
        <View>
          <TouchableOpacity
            activeOpacity={0.2}
            onPress={[Function]}
          >
            <AnimatedComponent
              accessible={true}
              isTVSelectable={true}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Array [
                  undefined,
                  Object {
                    "opacity": 1,
                  },
                ]
              }
            >
              <Component
                accessible={true}
                isTVSelectable={true}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                <View
                  accessible={true}
                  isTVSelectable={true}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Object {
                      "opacity": 1,
                    }
                  }
                >
                  <Component>
                    <Text>
                      undefined
                    </Text>
                  </Component>
                  <Component
                    source={
                      Object {
                        "uri": undefined,
                      }
                    }
                    style={
                      Object {
                        "height": 51,
                        "width": 51,
                      }
                    }
                  >
                    <Image
                      source={
                        Object {
                          "uri": undefined,
                        }
                      }
                      style={
                        Object {
                          "height": 51,
                          "width": 51,
                        }
                      }
                    />
                  </Component>
                </View>
              </Component>
            </AnimatedComponent>
          </TouchableOpacity>
        </View>
      </Component>
    </Article>
  </inject-Article-with-navigation>
</Provider>
`;

exports[`Article snapshot renders ArticleScreen 1`] = `
<ArticleScreen>
  <Component>
    <View>
      <TouchableOpacity
        activeOpacity={0.2}
        onPress={[Function]}
      >
        <AnimatedComponent
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              undefined,
              Object {
                "opacity": 1,
              },
            ]
          }
        >
          <Component
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <View
              accessible={true}
              isTVSelectable={true}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
            >
              <Component>
                <Text>
                  Back
                </Text>
              </Component>
            </View>
          </Component>
        </AnimatedComponent>
      </TouchableOpacity>
      <Component>
        <Text>
          ArticleScreen
        </Text>
      </Component>
    </View>
  </Component>
</ArticleScreen>
`;
